src/exchanges/binance/binance.exchange.ts

```diff
@@ -368,14 +371,65 @@ export class Binance extends BaseExchange {
   };
 
   fetchOHLCV = async (opts: OHLCVOptions) => {
-    const { data } = await this.xhr.get<any[][]>(ENDPOINTS.KLINE, {
-      params: {
-        symbol: opts.symbol,
-        interval: opts.interval,
-        limit: 500,
-      },
+    const [, amount, unit] = opts.interval.split(/(\d+)/);
+
+    // Default to 500
+    let requiredCandles = opts.limit ?? 500;
+
+    const startTime =
+      opts.startTime ??
+      dayjs()
+        .subtract(parseFloat(amount) * requiredCandles, unit as ManipulateType)
+        .unix();
+
+    // Calculate the number of candles that are going to be fetched
+    if (opts.endTime) {
+      const diff = dayjs
+        .unix(startTime)
+        .diff(dayjs.unix(opts.endTime), unit as ManipulateType);
+
+      requiredCandles = Math.abs(diff);
+    }
+
+    // Binance API only allows to fetch 1500 candles at a time
+    // so we need to split the request in multiple calls
+    const totalPages = Math.ceil(requiredCandles / KLINES_LIMIT);
+    const promises = [];
+    let currentStartTime = startTime;
+
+    for (let i = 0; i < totalPages; i++) {
+      const currentLimit = Math.min(
+        requiredCandles - i * KLINES_LIMIT,
+        KLINES_LIMIT
+      );
+      const currentEndTime = dayjs
+        .unix(currentStartTime)
+        .add(currentLimit, unit as ManipulateType)
+        .unix();
+
+      promises.push(
+        this.xhr.get<any[][]>(ENDPOINTS.KLINE, {
+          params: {
+            symbol: opts.symbol,
+            interval: opts.interval,
+            startTime: currentStartTime,
+            endTime: currentEndTime,
+            limit: currentLimit,
+          },
+        })
+      );
+      currentStartTime = currentEndTime;
+    }
+
+    const results = await Promise.all(promises);
+    const arr = results.flatMap((response) => {
+      const data = Array.isArray(response.data) ? response.data : [];
+      return data.filter((c: any) => c);
     });
 
+    // sort by timestamp and remove duplicated candles
+    const data = orderBy(uniqBy(arr, 0), 0, ['asc']);
+
```

Comment:
I've updated bit your implementation:

- using `mapSeries` to control the flow and request limits
- calculating from using pageIdx, which avoids using a `let`

```ts
const totalPages = Math.ceil(requiredCandles / KLINES_LIMIT);

    const results = await mapSeries(
      times(totalPages, (i) => i + 1),
      async (page) => {
        const currentLimit = Math.min(
          requiredCandles - page * KLINES_LIMIT,
          KLINES_LIMIT
        );

        const from = dayjs
          .unix(startTime)
          .add(currentLimit * page, unit as ManipulateType)
          .unix();

        const { data } = await this.xhr.get(ENDPOINTS.KLINE, {
          params: {
            symbol: opts.symbol,
            from,
            interval,
            limit: currentLimit,
          },
        });

        return data;
      }
    );

    const arr = results.flatMap((data) =>
      (Array.isArray(data.result) ? data.result : []).filter((c) => c)
    );

    const withoutDuplicates = uniqBy(arr, 'open_time');
    const ordered = orderBy(withoutDuplicates, ['open_time'], ['asc']);

    const candles: Candle[] = ordered.map((c: Record<string, any>) => {
```

Reply:
Done